tests/unit/s2n_self_talk_inet_socket_test.c

```diff
@@ -0,0 +1,163 @@
+
+    char sync = 0;
+    char recv_buffer[1] = { 0 };
+
+    {
+        RESULT_GUARD_POSIX(read(read_pipe, &sync, 1));
+        RESULT_GUARD_POSIX(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        RESULT_ENSURE_EQ(memcmp(&CHAR_A, &recv_buffer[0], 1), 0);
+
+        RESULT_GUARD_POSIX(read(read_pipe, &sync, 1));
+        RESULT_GUARD_POSIX(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        RESULT_ENSURE_EQ(memcmp(&CHAR_B, &recv_buffer[0], 1), 0);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT start_server(int fd, int write_pipe)
+{
+    /* Setup connections */
+    DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+            s2n_connection_ptr_free);
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Setup config */
+    EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+    EXPECT_EQUAL(s2n_connection_get_delay(server_conn), 0);
+    EXPECT_SUCCESS(s2n_connection_set_fd(server_conn, fd));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+    /* Do handshake */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    EXPECT_SUCCESS(s2n_negotiate(server_conn, &blocked));
+    EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+
+    char sync = 0;
+    char send_buffer[1] = { 0 };
+    {
+        send_buffer[0] = CHAR_A;
+        EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+        EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+
+        send_buffer[0] = CHAR_B;
+        EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+        EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    signal(SIGPIPE, SIG_IGN);
+
+    /* configure real socket */
+    int listener = socket(AF_INET, SOCK_STREAM, 0);
+    EXPECT_SUCCESS(listener);
+    struct sockaddr_in saddr = { 0 };
+    saddr.sin_family = AF_INET;
+    saddr.sin_addr.s_addr = htonl(S2N_TEST_INADDR_LOOPBACK);
+    saddr.sin_port = 0;
+
+    /* listen on socket address */
+    socklen_t addrlen = sizeof(saddr);
+    EXPECT_SUCCESS(bind(listener, (struct sockaddr *) &saddr, addrlen));
+    EXPECT_SUCCESS(getsockname(listener, (struct sockaddr *) &saddr, &addrlen));
+
+    /* used for synchronizing read and writes between client and server */
+    int sync_pipe[2] = { 0 };
+    EXPECT_SUCCESS(pipe(sync_pipe));
+
+    pid_t child = fork();
```

Comment:
I'm uncertain, but I thought we moved away from using explicit forks in our tests. -- Although I'm not sure what they got replaced with.

Reply:
They were replaced with [s2n_negotiate_test_server_and_client](https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L193) + [non-blocking io sockets or stuffers](https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L46-L67). You use forks if both sides need to be talking at the same time. If they're going to alternate talking, you don't need forks.

Reply:
Talked to Apoorv offline: he DOES need both sides talking at the same time for the listen/connect part. But I wonder if we can isolate just that part with threads or something so that we can otherwise keep everything in the same process?